File: libs/sdk/src/repair.ts
import { access, readFile } from 'node:fs/promises';
import { join } from 'node:path';

import {
  deriveMcpUrl,
  getConfigDir,
  type MoltNetConfig,
  writeConfig,
} from './credentials.js';

export interface ConfigIssue {
  field: string;
  problem: string;
  action: 'fixed' | 'warning';
}

export interface RepairResult {
  issues: ConfigIssue[];
  config: MoltNetConfig | null;
}

/**
 * Validate and optionally repair a MoltNet config.
 *
 * Checks required fields and detects stale file paths.
 *
 * Pass `dryRun: true` to report issues without writing changes.
 */
export async function repairConfig(opts?: {
  configDir?: string;
  dryRun?: boolean;
}): Promise<RepairResult> {
  const dir = opts?.configDir ?? getConfigDir();
  const issues: ConfigIssue[] = [];

  const config = await tryReadJson(join(dir, 'moltnet.json'));
  if (!config) {
    return { issues: [], config: null };
  }

  validateConfig(config, issues);
  await checkFilePaths(config, issues);

  // Apply auto-fixes
  if (!config.endpoints.mcp && config.endpoints.api) {
    config.endpoints.mcp = deriveMcpUrl(config.endpoints.api);
    issues.push({
      field: 'endpoints.mcp',
      problem: 'missing — derived from API endpoint',
      action: 'fixed',
    });
  }

  const hasAutoFixes = issues.some((i) => i.action === 'fixed');
  if (hasAutoFixes && !opts?.dryRun) {
    await writeConfig(config, dir);
  }

  return { issues, config };
}

function validateConfig(config: MoltNetConfig, issues: ConfigIssue[]): void {
  if (!config.identity_id) {
    issues.push({
      field: 'identity_id',
      problem: 'missing',
      action: 'warning',
    });
  }
  if (!config.keys.public_key) {
    issues.push({
      field: 'keys.public_key',
      problem: 'missing',
      action: 'warning',
    });
  }
  if (!config.keys.private_key) {
    issues.push({
      field: 'keys.private_key',
      problem: 'missing',
      action: 'warning',
    });
  }
  if (
    config.keys.public_key &&
    !config.keys.public_key.startsWith('ed25519:')
  ) {
    issues.push({
      field: 'keys.public_key',
      problem: "missing 'ed25519:' prefix",
      action: 'warning',
    });
  }
  if (!config.endpoints.api) {
    issues.push({
      field: 'endpoints.api',
      problem: 'missing',
      action: 'warning',
    });
  }
  if (config.github?.app_id && !/^\d+$/.test(config.github.app_id)) {
    issues.push({
      field: 'github.app_id',
      problem: `not a numeric GitHub App ID (got "${config.github.app_id}") — likely the slug; refetch via \`moltnet config init-from-env\` or re-run \`legreffier\``,
      action: 'warning',
    });
  }
}

async function checkFilePaths(
  config: MoltNetConfig,
  issues: ConfigIssue[],
): Promise<void> {
  const checks: { field: string; path: string }[] = [];

  if (config.ssh?.private_key_path) {
    checks.push({
      field: 'ssh.private_key_path',
      path: config.ssh.private_key_path,
    });
  }
  if (config.ssh?.public_key_path) {
    checks.push({
      field: 'ssh.public_key_path',
      path: config.ssh.public_key_path,
    });
  }
  if (config.git?.config_path) {
    checks.push({ field: 'git.config_path', path: config.git.config_path });
  }
  if (config.github?.private_key_path) {
    checks.push({
      field: 'github.private_key_path',
      path: config.github.private_key_path,
    });
  }

  for (const { field, path } of checks) {
    try {
      await access(path);
    } catch {
      issues.push({
        field,
        problem: `file not found: ${path}`,
        action: 'warning',
      });
    }
  }
}

async function tryReadJson(path: string): Promise<MoltNetConfig | null> {
  try {
    const content = await readFile(path, 'utf-8');
    return JSON.parse(content) as MoltNetConfig;
  } catch {
    return null;
  }
}

File: cmd/moltnet/repair.go
package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// ConfigIssue represents a single problem found during config validation.
type ConfigIssue struct {
	Field   string
	Problem string
	Action  string // "fixed", "warning", "migrate"
}

func runConfigRepair(args []string) error {
	fs := flag.NewFlagSet("config repair", flag.ExitOnError)
	credPath := fs.String("credentials", "", "Path to moltnet.json")
	dryRun := fs.Bool("dry-run", false, "Report issues without fixing them")

	fs.Usage = func() {
		fmt.Fprintln(os.Stderr, "Usage: moltnet config repair [options]")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "Validate and repair a MoltNet config file.")
		fmt.Fprintln(os.Stderr, "Checks required fields, fixes stale file paths,")
		fmt.Fprintln(os.Stderr, "and migrates credentials.json to moltnet.json.")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "Options:")
		fs.PrintDefaults()
	}

	if err := fs.Parse(args); err != nil {
		return err
	}

	resolvedPath, creds, issues, err := loadAndValidate(*credPath)
	if err != nil {
		return err
	}

	if len(issues) == 0 {
		fmt.Fprintln(os.Stderr, "Config is valid, no issues found.")
		return nil
	}

	fmt.Fprintf(os.Stderr, "Found %d issue(s):\n", len(issues))
	for _, iss := range issues {
		fmt.Fprintf(os.Stderr, "  [%s] %s: %s\n", iss.Action, iss.Field, iss.Problem)
	}

	if *dryRun {
		return nil
	}

	// Apply fixes
	fixed := 0
	for _, iss := range issues {
		if iss.Action == "migrate" {
			newPath, err := migrateConfig(resolvedPath, creds)
			if err != nil {
				return fmt.Errorf("migrate: %w", err)
			}
			resolvedPath = newPath
			fixed++
		}
		if iss.Action == "fixed" {
			fixed++
		}
	}

	if fixed > 0 {
		writePath := resolvedPath
		if *credPath != "" {
			writePath = *credPath
		}
		if _, err := WriteConfigTo(creds, writePath); err != nil {
			return fmt.Errorf("write config: %w", err)
		}
		fmt.Fprintf(os.Stderr, "\n%d issue(s) fixed. Config written to %s\n", fixed, writePath)
	}

	return nil
}

// loadAndValidate reads the config and returns all issues found.
// It mutates the config struct in-place for auto-fixable issues.
func loadAndValidate(credPath string) (string, *CredentialsFile, []ConfigIssue, error) {
	var issues []ConfigIssue
	var configPath string
	var creds *CredentialsFile

	if credPath != "" {
		configPath = credPath
		c, err := ReadConfigFrom(credPath)
		if err != nil {
			return "", nil, nil, fmt.Errorf("read config: %w", err)
		}
		if c == nil {
			return "", nil, nil, fmt.Errorf("config not found at %s", credPath)
		}
		creds = c
	} else {
		dir, err := GetConfigDir()
		if err != nil {
			return "", nil, nil, err
		}

		moltnetPath := filepath.Join(dir, "moltnet.json")
		legacyPath := filepath.Join(dir, "credentials.json")

		c, err := ReadConfigFrom(moltnetPath)
		if err != nil {
			return "", nil, nil, err
		}
		if c != nil {
			configPath = moltnetPath
			creds = c
		} else {
			c, err = ReadConfigFrom(legacyPath)
			if err != nil {
				return "", nil, nil, err
			}
			if c != nil {
				configPath = legacyPath
				creds = c
				issues = append(issues, ConfigIssue{
					Field:   "file",
					Problem: fmt.Sprintf("using deprecated credentials.json at %s — will migrate to moltnet.json", legacyPath),
					Action:  "migrate",
				})
			} else {
				return "", nil, nil, fmt.Errorf("no config found at %s or %s", moltnetPath, legacyPath)
			}
		}
	}

	// Required fields
	if creds.IdentityID == "" {
		issues = append(issues, ConfigIssue{Field: "identity_id", Problem: "missing", Action: "warning"})
	}
	if creds.Keys.PublicKey == "" {
		issues = append(issues, ConfigIssue{Field: "keys.public_key", Problem: "missing", Action: "warning"})
	}
	if creds.Keys.PrivateKey == "" {
		issues = append(issues, ConfigIssue{Field: "keys.private_key", Problem: "missing", Action: "warning"})
	}

	// Public key format
	if creds.Keys.PublicKey != "" && !strings.HasPrefix(creds.Keys.PublicKey, "ed25519:") {
		issues = append(issues, ConfigIssue{Field: "keys.public_key", Problem: "missing 'ed25519:' prefix", Action: "warning"})
	}

	// MCP endpoint derivable from API
	if creds.Endpoints.MCP == "" && creds.Endpoints.API != "" {
		creds.Endpoints.MCP = creds.Endpoints.API + "/mcp"
		issues = append(issues, ConfigIssue{Field: "endpoints.mcp", Problem: "missing — derived from API endpoint", Action: "fixed"})
	}
	if creds.Endpoints.API == "" {
		issues = append(issues, ConfigIssue{Field: "endpoints.api", Problem: "missing", Action: "warning"})
	}

	// File path validations
	if creds.SSH != nil {
		checkFilePath(&issues, "ssh.private_key_path", creds.SSH.PrivateKeyPath)
		checkFilePath(&issues, "ssh.public_key_path", creds.SSH.PublicKeyPath)
	}
	if creds.Git != nil {
		checkFilePath(&issues, "git.config_path", creds.Git.ConfigPath)
	}
	if creds.GitHub != nil {
		checkFilePath(&issues, "github.private_key_path", creds.GitHub.PrivateKeyPath)
	}

	return configPath, creds, issues, nil
}

func checkFilePath(issues *[]ConfigIssue, field, path string) {
	if path == "" {
		return
	}
	if _, err := os.Stat(path); os.IsNotExist(err) {
		*issues = append(*issues, ConfigIssue{
			Field:   field,
			Problem: fmt.Sprintf("file not found: %s", path),
			Action:  "warning",
		})
	}
}

// migrateConfig writes the config to moltnet.json in the same directory.
func migrateConfig(legacyPath string, creds *CredentialsFile) (string, error) {
	dir := filepath.Dir(legacyPath)
	newPath := filepath.Join(dir, "moltnet.json")

	if _, err := WriteConfigTo(creds, newPath); err != nil {
		return "", err
	}
	fmt.Fprintf(os.Stderr, "  Migrated to %s (you can now delete %s)\n", newPath, legacyPath)
	return newPath, nil
}
